File: leetcode/graph/sumRootToLeaf.ts
// Definition for a binary tree node.
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function sumNumbers(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  // Dfs
  const stack: {
    node: TreeNode;
    path: string;
  }[] = [];
  const visited = new Set();
  let totalPath: number[] = [];
  stack.push({
    node: root,
    path: `${root.val}`,
  });
  while (stack.length > 0) {
    const { node, path } = stack.pop() as { node: TreeNode; path: string };
    if (!node.left && !node.right) {
      totalPath.push(parseInt(`${path}`));
      continue;
    }
    if (node.left && !visited.has(node.left)) {
      stack.push({
        node: node.left,
        path: `${path}${node.left.val}`,
      });
      visited.add(node.left);
    }
    if (node.right && !visited.has(node.right)) {
      stack.push({
        node: node.right,
        path: `${path}${node.right.val}`,
      });
      visited.add(node.right);
    }
  }
  return totalPath.reduce((acc, cur) => acc + cur, 0);
}

File: leetcode/top150/binaryTree/sumRootToLeaf.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        stack = [(root, root.val)]
        total = 0
        while stack:
            node, num = stack.pop()
            if not node.left and not node.right:
                total += num
            if node.left:
                stack.append((node.left, num*10 + node.left.val))
            if node.right:
                stack.append((node.right, num*10 + node.right.val))
        return total